pkg/registry/core/pod/storage/eviction.go

```diff
@@ -104,40 +104,85 @@ func propagateDryRun(eviction *policy.Eviction, options *metav1.CreateOptions) (
 	eviction, ok := obj.(*policy.Eviction)
 	if !ok {
-		return nil, errors.NewBadRequest(fmt.Sprintf("not a Eviction object: %T", obj))
+		return nil, kerrors.NewBadRequest(fmt.Sprintf("not a Eviction object: %T", obj))
 	}
 
 	if name != eviction.Name {
-		return nil, errors.NewBadRequest("name in URL does not match name in Eviction object")
+		return nil, kerrors.NewBadRequest("name in URL does not match name in Eviction object")
 	}
 
-	deletionOptions, err := propagateDryRun(eviction, options)
+	originalDeleteOptions, err := propagateDryRun(eviction, options)
 	if err != nil {
 		return nil, err
 	}
 
-	obj, err = r.store.Get(ctx, eviction.Name, &metav1.GetOptions{})
-	if err != nil {
-		return nil, err
-	}
-	pod := obj.(*api.Pod)
-
 	if createValidation != nil {
 		if err := createValidation(ctx, eviction.DeepCopyObject()); err != nil {
 			return nil, err
 		}
 	}
 
-	// Evicting a terminal pod should result in direct deletion of pod as it already caused disruption by the time we are evicting.
-	// There is no need to check for pdb.
-	if pod.Status.Phase == api.PodSucceeded || pod.Status.Phase == api.PodFailed {
+	var pod *api.Pod
+	deletedPod := false
+	var unrecoverableConflict *error
+	err = retry.RetryOnConflict(EvictionsRetry, func() error {
+		obj, err = r.store.Get(ctx, eviction.Name, &metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		pod = obj.(*api.Pod)
+
+		// Evicting a terminal pod should result in direct deletion of pod as it already caused disruption by the time we are evicting.
+		// There is no need to check for pdb.
+		if !canIgnorePDB(pod) {
+			// Pod is not in a state where we can skip checking PDBs, exit the loop, and continue to PDB checks.
+			return nil
+		}
+
+		// the PDB can be ignored, so delete the pod
+		deletionOptions := originalDeleteOptions.DeepCopy()
+		// We should check if resourceVersion is already set by the requestor
+		// as it might be older than the pod we just fetched and should be
+		// honored.
+		if shouldEnforceResourceVersion(pod) && resourceVersionIsUnset(originalDeleteOptions) {
+			// Set deletionOptions.Preconditions.ResourceVersion to ensure we're not
+			// racing with another PDB-impacting process elsewhere.
+			if deletionOptions.Preconditions == nil {
+				deletionOptions.Preconditions = &metav1.Preconditions{}
+			}
+			deletionOptions.Preconditions.ResourceVersion = &pod.ResourceVersion
+		}
 		_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, deletionOptions)
 		if err != nil {
-			return nil, err
+			if kerrors.IsConflict(err) && !resourceVersionIsUnset(originalDeleteOptions) {
+				// We hit a conflict here for a requestor-specified resourceVersion.
+				// We save the error and return something else.
+				unrecoverableConflict = &err
+				return errors.New("Unrecoverable Conflict")
```

Comment:
Need to create a new error type here otherwise there's no way to break the retry loop that I'm aware of.  kerrors.IsConflict also works with wrapped errors according to docs, so that possibly won't work either.

Reply:
see suggestion at https://github.com/kubernetes/kubernetes/pull/83906#discussion_r432075989